src/vs/vscode.d.ts

```diff
@@ -1349,6 +1349,72 @@ declare module 'vscode' {
 	}
 
 	/**
+	 * A node provider for the tree explorer contributed by extension.
+	 *
+	 * Providers are registered through (#workspace.registerTreeExplorerNodeProvider) with a
+	 * `providerId` that corresponds to the `treeExplorerNodeProviderId` in the extension's
+	 * `contributes.explorer` section.
+	 *
+	 * The contributed tree explorer will ask the corresponding provider to provide the root
+	 * node and resolve children for each node. In addition, the provider could **optionally**
+	 * provide the following information for each node:
+	 * - label: A human-readable label used for rendering the node.
+	 * - hasChildren: Whether the node has children and is expandable.
+	 * - clickCommand: A command to execute when the node is clicked.
+	 */
+	export interface TreeExplorerNodeProvider<T> {
```

Comment:
Thanks for the feedback!

---

> Not a big friend of the lonely \<T\>

`TreeItem` was what I was initially using. However after talking to Kai and Daniel we feel the current approach is cleaner. The main advantages are:

- We wanted to allow `TreeItem` to be any shape. Just taking a few properties off the `TreeItem` for rendering feels bad.
- Separating concerns of storing data(`T`, extension will do that) and rendering (all logic in Provider).
- From extension author's POV, extending `TreeItem` and then upcasting the `TreeItem` to their own class in `resolveChildren` and the registered commands callback feels weird.
- The extension looks pretty good with TS2's [tagged union](https://github.com/Microsoft/TypeScript/wiki/What's-new-in-TypeScript#tagged-union-types): https://github.com/octref/explorer-api-example/blob/master/dep/src/extension.ts

---

> The node provider should have a way to signal to the main side that a node/substree has changes

Agreed. I also wanted this. It's not there as Kai was initially averse to it, but he now seems to support it. We have a refresh button on top of the viewlet. 

---

> Tree elements should support icons - in one or the other way

Decided to leave that for next iteration. Wanted to get some feedback from users before deciding how to do it.

---

> We should consider adding something like paging to the resolveChildren-call. I see this API being used for large sets or sets that are slow to resolve.

We are showing the progress bar if the `resolveChildren` is taking long, similar to what the search viewlet is doing. I feel we can delegate paging to extension authors. Resolve first page -> notify VSCode of change in the subtree when they resolve next page.

Better to wait for some feedback before adding it, I think.

---

> The API lacks a way to reveal, select, focus a node in a tree.

Yep. But while I was writing my 3 examples I never found the need for it. And I can't think of a use case for DB Explorer, which we are designing this for.

So better wait for next iteration when we understand the use cases more.

---

Will sync with Kai/Daniel this afternoon and talk through your concerns.